### PR TITLE
Fix function_clause exception on invalid DB security objects

### PR DESCRIPTION
**Pull request by wenli200133**
## Overview

 This PR will fix function_clause error on invalid DB security objects.

for example when the request body of PUT db/_security endpoint is not a correct json format, if you check the log file you will find the function_clause error.
```
[error] 2018-06-13T16:50:50.479454Z node1@127.0.0.1 <0.3642.0> 3108587a16 rpc couch_db:set_security/2 function_clause [{couch_db,validate_names_and_roles,[[<<"role_a">>]],[{file,"src/couch_db.erl"},{line,594}]},{couch_db,validate_security_object,1,[{file,"src/couch_db.erl"},{line,590}]},{couch_db,set_security,2,[{file,"src/couch_db.erl"},{line,574}]},{fabric_rpc,with_db,3,[{file,"src/fabric_rpc.erl"},{line,289}]},{rexi_server,init_p,3,[{file,"src/rexi_server.erl"},{line,139}]}]                                                                              [error] 2018-06-13T16:50:50.479648Z node1@127.0.0.1 <0.3576.0> 3108587a16 Failed to set security object on {'node1@127.0.0.1',<<"shards/40000000-5fffffff/abc.1528908305">>} :: {error,function_clause}                                         [error] 2018-06-13T16:50:50.480220Z node1@127.0.0.1 <0.3647.0> 3108587a16 rpc couch_db:set_security/2 function_clause [{couch_db,validate_names_and_roles,[[<<"role_a">>]],[{file,"src/couch_db.erl"},{line,594}]},{couch_db,validate_security_object,1,[{file,"src/couch_db.erl"},{line,590}]},{couch_db,set_security,2,[{file,"src/couch_db.erl"},{line,574}]},{fabric_rpc,with_db,3,[{file,"src/fabric_rpc.erl"},{line,289}]},{rexi_server,init_p,3,[{file,"src/rexi_server.erl"},{line,139}]}]                                                                              [error] 2018-06-13T16:50:50.480371Z node1@127.0.0.1 <0.3646.0> 3108587a16 rpc couch_db:set_security/2 function_clause [
{couch_db,validate_names_and_roles,[[<<"role_a">>]],[{file,"src/couch_db.erl"},{line,594}]},
{couch_db,validate_security_object,1,[{file,"src/couch_db.erl"},{line,590}]},
{couch_db,set_security,2,[{file,"src/couch_db.erl"},{line,574}]},
{fabric_rpc,with_db,3,[{file,"src/fabric_rpc.erl"},{line,289}]},
{rexi_server,init_p,3,[{file,"src/rexi_server.erl"},{line,139}]}] 
```
after fix it,` function_clause` error full stack trace will disappeared instead of
```
[error] 2018-11-29T02:30:24.013166Z node1@127.0.0.1 <0.2699.0> 985a72f160 Failed to set security object on {'node1@127.0.0.1',<<"shards/e0000000-ffffffff/test.1543202135">>} :: "admins or members must be a JSON list of strings"
```

## Testing recommendations

case 

1. test with incorrect members format
`curl -i -X PUT http://localhost:5984/test/_security -u "adm:pass" -H 'accept: application/json'  -d '{"members":["foo"]}'`
check logs:
```
node2.log:8979:[error] 2018-11-29T14:08:05.480944Z node2@127.0.0.1 <0.550.0> 71a87732c7 Failed to set security object on {'node3@127.0.0.1',<<"shards/c0000000-dfffffff/test.1543202135">>} :: "admins or members must be a JSON list of strings"
```
return 
`{"error":"error","reason":"no_majority"}`

2.  test with incorrect members roles format

`curl -i -X PUT http://localhost:15984/test/_security -u "adm:pass" -H 'accept: application/json'  -d '{"members":{"names": ["user1","user2"],"roles": "developers"}}'`

check log:
```
node1.log:19740:[error] 2018-11-29T14:03:44.740886Z node1@127.0.0.1 <0.5831.0> 76fbe3cd55 Failed to set security object on {'node2@127.0.0.1',<<"shards/e0000000-ffffffff/test.1543202135">>} :: "roles field vallue must be a list of strings"
```
return 
`{"error":"error","reason":"no_majority"}`

3. test with correct members roles and names format
`curl -i -X PUT http://localhost:5984/test/_security -u "adm:pass" -H 'accept: application/json'  -d '{"members":{"names": ["user1","user2"],"roles": ["developers"]}}'`

return 
`{"ok":true}`

4. test with correct admins and members roles and names format
`curl -i -X PUT http://localhost:5984/test/_security -u "adm:pass" -H 'accept: application/json'  -d '{"admins":{"roles":["admins"],"names":["foo"]},"members":{"names": ["user1","user2"],"roles": ["developers"]}}'`

return 
`{"ok":true}`

## Related Issues or Pull Requests

#1384 

## Checklist

- [x] Code is written and works correctly;
- [x] Changes are covered by tests;
- [ ] Documentation reflects the changes;

### CODE REVIEW

**Comment by eiri** [comment]
One more thing - can you please change your commit message (and PR titile) to something more descriptive? It'll read really cryptic to people out of context when merged. Something like a title of the original [issue](#1384) - "Fix function_clause exception on invalid DB security objects" or similar?

**Comment by wenli200133** [comment]
Hi @eiri Could you help me review again if you have a free time, I updated the codes according to your comments, thank you for your time.

**Comment by jiangphcn** [comment]
generally, looks better after adding @eiri 's comments. Just minor comments, there are still some lines longer than 80 characters.

File: src/chttpd/test/chttpd_security_tests.erl [hunk 1]
```diff
@@ -126,7 +147,6 @@ should_allow_admin_db_compaction([Url,_UsersUrl]) ->
             couch_util:get_value(<<"ok">>, InnerJson, undefined)
         end).
 
-
```
Comment (eiri) [inline src/chttpd/test/chttpd_security_tests.erl]: Empty line changes should be in a separate commit

File: src/chttpd/test/chttpd_security_tests.erl [hunk 2]
```diff
@@ -207,3 +227,93 @@ should_disallow_anonymous_db_view_cleanup([Url,_UsersUrl]) ->
...
+    SecurityUrl = lists:concat([Url, "/_security"]),
+    SecurityProperties = [
+        {<<"admins">>,{[{<<"roles">>,[?TEST_ADMIN]}]}},
+        {<<"members">>,{[{<<"roles">>,[<<?TEST_MEMBER>>]}]}}
+    ],
+
+    Body = jiffy:encode({SecurityProperties}),
+    {ok, Status, _, RespBody} = test_request:put(SecurityUrl,
+        [?CONTENT_JSON, ?AUTH], Body),
+    ?_assertEqual(500, Status),
```
Comment (eiri) [inline src/chttpd/test/chttpd_security_tests.erl]: Almost there, but still not correct. As I said, the multiple assertions must be returned as a list.

What is happening here: `?_assertEqual` doesn't run check, it returns function, that runs the assertion when called. So when you are calling `?_assert...` macro in a middle of callback test function, they not ran, but just ignored.

So what this function should return is this:
```erlang
[
    ?_assertEqual(500, Status),
    ?_assertEqual([
        {<<"error">>,<<"error">>},
        {<<"reason">>,<<"no_majority">>}
    ], ResultJson)
].
```

File: src/chttpd/test/chttpd_security_tests.erl [hunk 3]
```diff
@@ -207,3 +227,93 @@ should_disallow_anonymous_db_view_cleanup([Url,_UsersUrl]) ->
...
+    {ResultJson} = ?JSON_DECODE(RespBody),
+    ?_assertMatch([
```
Comment (eiri) [inline src/chttpd/test/chttpd_security_tests.erl]: As I said already, `?_assertMatch` is for guards assertion, and here we are asserting equality, so `?_assertEqual` more appropriate.

File: src/chttpd/test/chttpd_security_tests.erl [hunk 4]
```diff
@@ -207,3 +227,93 @@ should_disallow_anonymous_db_view_cleanup([Url,_UsersUrl]) ->
...
+    SecurityProperties = [
+        {<<"admins">>,{[{<<"roles">>,[?TEST_ADMIN]}]}},
+        {<<"members">>,{[{<<"roles">>,[<<?TEST_MEMBER>>]}]}}
+    ],
+
+    Body = jiffy:encode({SecurityProperties}),
+    {ok, Status, _, RespBody} = test_request:put(SecurityUrl,
+        [?CONTENT_JSON, ?AUTH], Body),
+    ?_assertEqual(500, Status),
+    {ResultJson} = ?JSON_DECODE(RespBody),
```
Comment (eiri) [inline src/chttpd/test/chttpd_security_tests.erl]: Don't do match here, return object and assert it later, i.e. `ResultJson = ?JSON_DECODE(RespBody)`. That way tests will return readable failure, instead of crashing with critic `no_match` exception as they'd do now.

File: src/couch/src/couch_db.erl [hunk 5]
```diff
@@ -736,19 +736,23 @@ validate_security_object(SecProps) ->
...
 % validate user input
 validate_names_and_roles({Props}) when is_list(Props) ->
-    case couch_util:get_value(<<"names">>,Props,[]) of
+    case couch_util:get_value(<<"names">>, Props, []) of
     Ns when is_list(Ns) ->
             [throw("names must be a JSON list of strings") ||N <- Ns, not is_binary(N)],
             Ns;
-    _ -> throw("names must be a JSON list of strings")
+    _ ->
+        throw("names field vallue must be a list of strings")
```
Comment (eiri) [inline src/couch/src/couch_db.erl]: typo in "vallue".

why change of error message, btw? I think it's clear as it is

File: src/couch/src/couch_db.erl [hunk 6]
```diff
@@ -736,19 +736,23 @@ validate_security_object(SecProps) ->
...
     end,
-    case couch_util:get_value(<<"roles">>,Props,[]) of
+    case couch_util:get_value(<<"roles">>, Props, []) of
     Rs when is_list(Rs) ->
         [throw("roles must be a JSON list of strings") ||R <- Rs, not is_binary(R)],
         Rs;
-    _ -> throw("roles must be a JSON list of strings")
+    _ ->
+        throw("roles field vallue must be a list of strings")
```
Comment (eiri) [inline src/couch/src/couch_db.erl]: same typo here

File: src/couch/src/couch_db.erl (hunk 5 above)
Comment (wenli200133) [reply src/couch/src/couch_db.erl]: here if request body is like 
`'{"members":{"names": "user1"}}'`
the logical will enter into line 744, throw an exception, but here key of `names` existed just value is not correct, the value should like this `["user1"]`, so I update the description.
but you think this is clear `names must be a JSON list of strings`, I can change it back.
to be honest I don't quiet know what it  a `JSON list of strings` look like.

Comment (eiri) [reply src/couch/src/couch_db.erl]: JSON list or array is an ordered list of zero or more values represented by square brackets. JSON string is a sequence of zero or more unicode characters delimited with double quotes.

So valid JSON list of strings is something that looks like `[]` or `["user1"]` or `["user1", "user2"]`. Since a security object has to be represent as a valid JSON, I think it's appropriate to refer to JSON when security object's content is invalid.

**Comment by wenli200133** [comment]
Hi @eiri could you help review again? thanks.

**Comment by wohali** [comment]
@wenli200133 Please edit the description so it says "Closes #1384" so the ticket auto-closes when this merges. Thank you for a great PR!

@jiangphcn You can go ahead and merge this on @wenli200133 's behalf. Be sure #1384 gets closed.

**Comment by jiangphcn** [comment]
@wohali sure, just rebased and merged this PR. Looks that #1384 is auto-closed.

@wenli200133 thanks for your great work.
